src/profile-logic/profile-data.js

```diff
@@ -455,14 +449,91 @@ export function getTimingsForCallNodeIndex(
       breakdownByCategory: null,
     },
   };
+
+  // This holds the root time, it's incremented for all samples and is useful to
+  // have an absolute value to compare the other values with.
   let rootTime = 0;
 
-  if (needleNodeIndex === null) {
-    // No index was provided, return empty timing information.
-    return { forPath: pathTimings, forFunc: funcTimings, rootTime };
+  /* -------- End of variable definitions ------- */
+
+  /* ------------ Functions definitions --------- *
+   * We define functions here so that they have easy access to the variables and
+   * the algorithm's parameters. */
+
+  /**
+   * This function is called for native stacks. If the native stack has the
+   * 'JavaScript' category, then we move up the call tree to find the nearest
+   * ancestor that's JS and returns its JS implementation.
+   */
+  function getImplementationForNativeStack(
+    unfilteredStackIndex: IndexIntoStackTable
+  ): StackImplementation {
+    const category = unfilteredStackTable.category[unfilteredStackIndex];
+    if (category !== javascriptCategoryIndex) {
+      return 'native';
+    }
+
+    for (
+      let currentStackIndex = unfilteredStackIndex;
+      currentStackIndex !== null;
+      currentStackIndex = unfilteredStackTable.prefix[currentStackIndex]
+    ) {
+      const frameIndex = unfilteredStackTable.frame[currentStackIndex];
+      const funcIndex = unfilteredFrameTable.func[frameIndex];
+      const isJS = unfilteredFuncTable.isJS[funcIndex];
+      if (isJS) {
+        return getImplementationForJsStack(frameIndex);
+      }
+    }
+
+    // No JS frame was found in the ancestors, this is weird but why not?
+    return 'native';
   }
 
-  const needleFuncIndex = callNodeTable.func[needleNodeIndex];
+  /**
+   * This function Returns the JS implementation information for a specific JS stack.
+   */
+  function getImplementationForJsStack(
+    unfilteredFrameIndex: IndexIntoFrameTable
+  ): JsImplementation {
+    const jsImplementationStrIndex =
+      unfilteredFrameTable.implementation[unfilteredFrameIndex];
+
+    if (jsImplementationStrIndex === null) {
+      return 'interpreter';
+    }
+
+    const jsImplementation = stringTable.getString(jsImplementationStrIndex);
+
+    switch (jsImplementation) {
+      case 'baseline':
+      case 'blinterp':
+      case 'ion':
+        return jsImplementation;
+      default:
```

Comment:
Should this use an exhaustive check?

Reply:
It's not possible because `jsImplementation` is a simple string here :-) this function is especially to convert from a string to the enum.